tests/test_simple_nlp.py

```diff
@@ -125,3 +126,67 @@ def test_simple_nlp_mlm(tiny_bert_tokenizer, tiny_bert_model):
     num_predict_batches_expected = ((size - 1) // batch_size) + 1
     assert len(predictions) == num_predict_batches_expected
     assert predictions[0].shape == (batch_size, sequence_length, vocab_size)
+
+
+def test_simple_nlp_mlm_token_batch(tiny_bert_tokenizer, tiny_bert_model):
+    transformers = pytest.importorskip('transformers')
+
+    vocab_size = tiny_bert_tokenizer.vocab_size
+    sequence_length = 32
+    size = 96
+    batch_size = 8
+
+    train_dataset = RandomTextLMDataset(
+        size=size,
+        vocab_size=vocab_size,
+        sequence_length=sequence_length,
+        use_keys=True,
+        pad_token_id=tiny_bert_tokenizer.pad_token_id,
+    )
+    collator = transformers.DataCollatorForLanguageModeling(tokenizer=tiny_bert_tokenizer, mlm_probability=0.15)
+
+    model = SimpleTransformerMaskedLM(vocab_size=vocab_size)
+    state_dict = model.state_dict()
+
+    # Set up an ordinary trainer and get the model's state dict before training starts
+    train_dataloader = DataLoader(
+        train_dataset,
+        batch_size=batch_size,
+        sampler=dist.get_sampler(train_dataset),
+        collate_fn=collator,
+    )
+
+    trainer = Trainer(
+        model=model,
+        train_dataloader=train_dataloader,
+        max_duration='2ep',
+        accumulate_train_batch_on_tokens=False,
+    )
+
+    trainer.fit()
+
+    # Check that there is some train cross entropy
+    assert trainer.state.train_metrics is not None
+    cross_entropy = trainer.state.train_metrics['LanguageCrossEntropy'].compute()
+    assert cross_entropy != 0.0
+
+    # Set up a trainer that accumulates train loss based on token counts, after reloading original state dict
+    model.load_state_dict(state_dict)
+    token_data_spec = DataSpec(
+        dataloader=train_dataloader,
+        get_num_tokens_in_batch=lambda b: (b['input_ids'] != tiny_bert_tokenizer.pad_token_id).sum().item(),
+    )
+    token_trainer = Trainer(
+        model=model,
+        train_dataloader=token_data_spec,
+        accumulate_train_batch_on_tokens=False,
```

Comment:
I'm guessing you meant to make this True? And if this test passed as easy I'm guessing we aren't being quite strict enough here and the CEs are randomly different. Trying to think if there is any way we can test more directly than just seeing if the CE is different...

Reply:
Thanks for catching this. I think the problem is that I'm not controlling the randomness in the data collator. To work around this problem, let me
1. Proactively initialize the training dataset
2. Add `reproducibility.seed_all(42)` before each call to `trainer.fit()`
3. Add an extra sanity check that I can reproduce the CE calculation from the original regular trainer